### DISCUSSION TITLE
self bidirectional one-one relationship

### DISCUSSION THREAD

**Discussion by shaozi** [sqlalchemy/sqlalchemy#7385] [Usage Questions] [answered]
Hi,

I have tried a lot of online examples but still could not find a working one for my case. I have a table represents device ports, each port can be connect to another port, it is a one-to-one relationship. I want to use a single var called `peer` to represent this link relationship.

```python
from sqlalchemy.orm import relationship
from sqlalchemy.sql.schema import Column, ForeignKey
from cpr.database import db

from sqlalchemy.sql.sqltypes import Integer, String


class Port(db.Base):
    __tablename__ = 'port'
    id = Column(Integer, primary_key=True)
    name = Column(String)
    peer_id = Column(Integer, ForeignKey('port.id'))
    peer = relationship('Port', back_populates='peer', uselist=False, remote_side=[id], post_update=True)


port1 = Port(name='p1')
port2 = Port(name='p2')
port1.peer = port2
assert port1.peer.name == 'p2'
assert port2.peer.name == 'p1'
```

It either complain the `remote_side` should be on the many-to-one side, which in my case, it is the same as one-to-many side, or I have to remove the `back_populates`, which peer is not automatically set on the reverse direction.

Is there a way to do it?

**Comment by zzzeek** [comment, marked as answer]
short answer, this isn't supported.    you would need to make two different relationships, then use a `@property` or other descriptor to define "peer" as "self.left_peer or self.right_peer".

this came up awhile ago and after I failed to notice what they were doing for a bit I wrote a bunch about this use case here: https://github.com/sqlalchemy/sqlalchemy/issues/6137#issuecomment-808412223   - however, in that case, they are using a "secondary" table which makes it slightly more feasible, but IMO still too unusual and different of a use case to be worth supporting.

**Reply by zzzeek** [reply]
actually since you are sharing the same foreign key column this might not be possible at all.  i think the design would be cleaner if you had a "secondary" table that defines "peer-to-peer".

**Reply by zzzeek** [reply]
OK I think just one relationship and use a property, this is a POC that writes the data in the way you are looking for

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
Base = declarative_base()


class Port(Base):
    __tablename__ = 'port'
    id = Column(Integer, primary_key=True)
    name = Column(String)
    peer_id = Column(Integer, ForeignKey('port.id'))
    _peer = relationship('Port', uselist=False, remote_side=[id], post_update=True)

    @property
    def peer(self):
        return self._peer

    @peer.setter
    def peer(self, other):
        self._peer = other
        other._peer = self

port1 = Port(name='p1')
port2 = Port(name='p2')
port1.peer = port2

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)
s.add_all([port1, port2])
s.commit()

assert port1.peer.name == 'p2'
assert port2.peer.name == 'p1'
```

**Reply by shaozi** [reply]
This is so cool! 
I add this to the setter:
```python
    @peer.setter
    def peer(self, other):
        if self._peer:
            # disconnect the other end first
            self._peer._peer = None
        self._peer = other
        other._peer = self
```
purpose is if I do this:
```
port1.peer = port2
port1.peer = port3
```
then port1 will be only connected to port3, port2.peer will be None.

Do you see any problem with that?

Thanks,

**Reply by zzzeek** [reply]
should be fine, write a suite of test cases that include round-tripping to SQLite and you can make sure everything works as desired.

**Reply by shaozi** [reply]
Should consider the other side too:
```python
@peer.setter
    def peer(self, other):
        if self._peer:
            # disconnect the other end first
            self._peer._peer = None
        if other._peer:
            # disconnect the other's peer too
            other._peer._peer = None
        self._peer = other
        other._peer = self

```
